File: go/gdl/objects/interaction/component/component.go
package component

import (
	"encoding/json"
	"errors"
	"fmt"
)

type ComponentType uint8

const (
	ComponentActionRow ComponentType = iota + 1
	ComponentButton
	ComponentSelectMenu
	ComponentInputText
	ComponentUserSelect
	ComponentRoleSelect
	ComponentMentionableSelect
	ComponentChannelSelect
	ComponentSection
	ComponentTextDisplay
	ComponentThumbnail
	ComponentMediaGallery
	ComponentFile
	ComponentSeparator
	ComponentContainer ComponentType = iota + 3 // 14
	ComponentLabel
	ComponentFileUpload
	ComponentRadioGroup ComponentType = iota + 4 // 21
	ComponentCheckboxGroup
	ComponentCheckbox
)

type Component struct {
	Type ComponentType `json:"type"`
	ComponentData
}

type ComponentData interface {
	Type() ComponentType
}

var (
	ErrMissingType  = errors.New("component was missing type field")
	ErrUnknownType  = errors.New("component had unknown type")
	ErrTypeMismatch = errors.New("data did not match component type")
)

func (c Component) MarshalJSON() ([]byte, error) {
	return encode(c.ComponentData)
}

func encode(c ComponentData) (json.RawMessage, error) {
	switch v := c.(type) {
	case ActionRow:
		subComponents := make([]json.RawMessage, len(v.Components))
		for i, sub := range v.Components {
			var err error
			subComponents[i], err = encode(sub.ComponentData)
			if err != nil {
				return nil, err
			}
		}

		data := map[string]interface{}{
			"type":       ComponentActionRow,
			"components": subComponents,
		}
		return json.Marshal(data)
	case Button:
		return json.Marshal(v)
	case SelectMenu:
		return json.Marshal(v)
	case InputText:
		return json.Marshal(v)
	case UserSelect:
		return json.Marshal(v)
	case RoleSelect:
		return json.Marshal(v)
	case MentionableSelect:
		return json.Marshal(v)
	case ChannelSelect:
		return json.Marshal(v)
	case Section:
		return json.Marshal(v)
	case TextDisplay:
		return json.Marshal(v)
	case Thumbnail:
		return json.Marshal(v)
	case MediaGallery:
		return json.Marshal(v)
	case File:
		return json.Marshal(v)
	case Separator:
		return json.Marshal(v)
	case Container:
		return json.Marshal(v)
	case Label:
		return json.Marshal(v)
	case FileUpload:
		return json.Marshal(v)
	case RadioGroup:
		return json.Marshal(v)
	case CheckboxGroup:
		return json.Marshal(v)
	case Checkbox:
		return json.Marshal(v)
	default:
		fmt.Println(v)
		return nil, ErrUnknownType
	}
}

func (c *Component) UnmarshalJSON(data []byte) error {
	var raw map[string]interface{}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}

	var componentType ComponentType
	if rawType, ok := raw["type"].(float64); ok {
		componentType = ComponentType(rawType)
	} else {
		return ErrMissingType
	}

	var err error
	switch componentType {
	case ComponentActionRow:
		var parsed ActionRow
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentButton:
		var parsed Button
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentSelectMenu:
		var parsed SelectMenu
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentInputText:
		var parsed InputText
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentUserSelect:
		var parsed UserSelect
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentRoleSelect:
		var parsed RoleSelect
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentMentionableSelect:
		var parsed MentionableSelect
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentChannelSelect:
		var parsed ChannelSelect
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentSection:
		var parsed Section
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentTextDisplay:
		var parsed TextDisplay
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentThumbnail:
		var parsed Thumbnail
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentMediaGallery:
		var parsed MediaGallery
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentFile:
		var parsed File
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentSeparator:
		var parsed Separator
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentContainer:
		var parsed Container
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentLabel:
		var parsed Label
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentFileUpload:
		var parsed FileUpload
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentRadioGroup:
		var parsed RadioGroup
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentCheckboxGroup:
		var parsed CheckboxGroup
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	case ComponentCheckbox:
		var parsed Checkbox
		err = json.Unmarshal(data, &parsed)
		c.ComponentData = parsed
	default:
		return errors.Join(ErrUnknownType, fmt.Errorf("unknown component type %d", componentType))
	}

	if err != nil {
		return err
	}

	c.Type = componentType
	return nil
}

File: rust/model/src/interaction/component.rs
use super::{
    ActionRow, Button, CheckboxGroup, Container, File, InputText, Label, MediaGallery,
    RadioGroup, Section, SelectMenu, Separator, TextDisplay, Thumbnail,
};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Component {
    ActionRow(Box<ActionRow>),
    Button(Box<Button>),
    SelectMenu(Box<SelectMenu>),
    InputText(Box<InputText>),
    Section(Box<Section>),           // 9
    TextDisplay(Box<TextDisplay>),   // 10
    Thumbnail(Box<Thumbnail>),       // 11
    MediaGallery(Box<MediaGallery>), // 12
    File(Box<File>),                 // 13
    Separator(Box<Separator>),       // 14
    // 15 & 16 are not used
    Container(Box<Container>),       // 17
    Label(Box<Label>),               // 18
    // 19 & 20 are not used
    RadioGroup(Box<RadioGroup>),       // 21
    CheckboxGroup(Box<CheckboxGroup>), // 22
}

#[derive(Serialize_repr, Deserialize_repr, Debug, Copy, Clone)]
#[repr(u8)]
pub enum ComponentType {
    ActionRow = 1,
    Button = 2,
    SelectMenu = 3,
    InputText = 4,
    UserSelect = 5,
    RoleSelect = 6,
    MentionableSelect = 7,
    ChannelSelect = 8,
    Section = 9,
    TextDisplay = 10,
    Thumbnail = 11,
    MediaGallery = 12,
    File = 13,
    Separator = 14,
    // 15 & 16 are not used
    Container = 17,
    Label = 18,
    // 19 & 20 are not used
    RadioGroup = 21,
    CheckboxGroup = 22,
}

impl TryFrom<u64> for ComponentType {
    type Error = Box<str>;

    fn try_from(value: u64) -> Result<Self, Self::Error> {
        Ok(match value {
            1 => Self::ActionRow,
            2 => Self::Button,
            3 => Self::SelectMenu,
            4 => Self::InputText,
            5 => Self::UserSelect,
            6 => Self::RoleSelect,
            7 => Self::MentionableSelect,
            8 => Self::ChannelSelect,
            9 => Self::Section,
            10 => Self::TextDisplay,
            11 => Self::Thumbnail,
            12 => Self::MediaGallery,
            13 => Self::File,
            14 => Self::Separator,
            // 15 & 16 are not used
            17 => Self::Container,
            18 => Self::Label,
            // 19 & 20 are not used
            21 => Self::RadioGroup,
            22 => Self::CheckboxGroup,
            _ => return Err(format!("invalid component type \"{}\"", value).into_boxed_str()),
        })
    }
}

impl<'de> Deserialize<'de> for Component {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let value = Value::deserialize(deserializer)?;

        let component_type = value
            .get("type")
            .and_then(Value::as_u64)
            .ok_or_else(|| Box::from("component type was not an integer"))
            .and_then(ComponentType::try_from)
            .map_err(D::Error::custom)?;

        let component = match component_type {
            ComponentType::ActionRow => serde_json::from_value(value).map(Component::ActionRow),
            ComponentType::Button => serde_json::from_value(value).map(Component::Button),
            ComponentType::Section => serde_json::from_value(value).map(Component::Section),
            ComponentType::TextDisplay => serde_json::from_value(value).map(Component::TextDisplay),
            ComponentType::Thumbnail => serde_json::from_value(value).map(Component::Thumbnail),
            ComponentType::MediaGallery => {
                serde_json::from_value(value).map(Component::MediaGallery)
            }
            ComponentType::File => serde_json::from_value(value).map(Component::File),
            ComponentType::Separator => serde_json::from_value(value).map(Component::Separator),
            ComponentType::Container => serde_json::from_value(value).map(Component::Container),
            ComponentType::Label => serde_json::from_value(value).map(Component::Label),
            ComponentType::SelectMenu
            | ComponentType::UserSelect
            | ComponentType::RoleSelect
            | ComponentType::MentionableSelect
            | ComponentType::ChannelSelect => {
                serde_json::from_value(value).map(Component::SelectMenu)
            }
            ComponentType::InputText => serde_json::from_value(value).map(Component::InputText),
            ComponentType::RadioGroup => {
                serde_json::from_value(value).map(Component::RadioGroup)
            }
            ComponentType::CheckboxGroup => {
                serde_json::from_value(value).map(Component::CheckboxGroup)
            }
        }
        .map_err(D::Error::custom)?;

        Ok(component)
    }
}
